### ISSUE TITLE
no-multi-spaces does not allow non-last function param to be commented-out

### ISSUE THREAD

**Issue by jrencz** [eslint/eslint#1659]
Consider the following example:

```
function logAndCompare(a, b, c) {
    console.log(c);
    return a === b;
}
```
_listing 1.1 all params are used_

this is valid.
But if I want to temporarily comment-out the use of `c`, like this:

```
function logAndCompare(a, b, c) {
//    console.log(c);
    return a === b;
}
```
_listing 1.2: usage of the last param commented out_

I get noticed, that there's an unused param `c` and i get noticed by `no-unused-vars`. Reasonable.
Then I comment out the last param like this:

```
function logAndCompare(a, b/*, c*/) {
//    console.log(c);
    return a === b;
}
```
_listing 1.3: declaration of the last param commented out_

so I will remember what it was when I come back some days later to bring it back.

If I don't need `a` or `b` I will not get noticed by `no-unused-vars` since generally in JS the position of param does matter and it's required to define both `a` and `b` to be able to use `c`
If I ever wanted to comment out `b` the order of params will not be the same and when my function gets called `c` is now `b`.

For the AngularJS environment however position of function params is not as important as their names. With this in mind let's say my function is a service (i'm not using the annotated form on purpose - there's [ngAnnotate](https://github.com/olov/ng-annotate) for that):

```
angular.module('foo').service('fooService', function (aService, bService, cService) {
    aService.doAStuff();
    bService.doBStuff();
    cService.doCStuff();
})
```
_listing 2.1: all injected services used_

If I want to comment out `aService` I'm aware that 'no-unused-vars' won't warn me:

```
angular.module('foo').service('fooService', function (aService, bService, cService) {
//    aService.doAStuff();
    bService.doBStuff();
    cService.doCStuff();
})
```
_listing 2.2: not all injected services used_

What I'd like to do (to better emphasise that I'm aware service `aService` is no longer used but it's most probable it will be back) is to do this:

```
angular.module('foo').service('fooService', function (/*aService, */bService, cService) {
//    aService.doAStuff();
    bService.doBStuff();
    cService.doCStuff();
})
```
_listing 2.3: declaration of `aService` removed._

IT does not look like a good idea but if my `fooService` has many services with long names injected and I want to both have my diffs as small as they can be and have those injected services in a particular order but not exceed the limit of line length I get stuck.

```
angular.module('foo').service('fooService', function (aServiceWithLongerName, bServiceWithLongerName, 
                                                      cService) {
    aService.doAStuff();
    bService.doBStuff();
    cService.doCStuff();
})
```
_listing 2.4: multiline induced by line length limit_

```
angular.module('foo').service('fooService', function (bServiceWithLongerName, cService) {
//    aService.doAStuff();
    bService.doBStuff();
    cService.doCStuff();
})
```
_listing 2.4: service removed: longer and less meaningful diff_

In terms of "Vanilla" JS this gets all messed up - I know. But in AngularJS this is perfectly valid and can even be considered a good practice.
I cannot do this though because `no-multi-spaces` fails.

I created 3 cases: https://github.com/jrencz/eslint/blob/jrencz/issue/comment/tests/lib/rules/no-multi-spaces.js#L49-51
Only the last one passes the tests. I suppose this is a bug and all of them are valid in terms of syntax but such case only makes sense when AngularJS is used (or the order does not matter for some other reason)

**Comment by nzakas** [comment]
Thanks for the details - can you summarize the change you're looking for? It's not clear from your description.

Also, if you could provide an example of the code plus the output you see from ESLint that you think is incorrect, that would help a lot.

**Comment by jrencz** [comment]
Please check out the branch on my fork and run tests.

https://github.com/jrencz/eslint/blob/jrencz/issue/comment/tests/lib/rules/no-multi-spaces.js#L49-51

**Comment by nzakas** [comment]
I'll be happy to do that. I'd still appreciate it if you could provide the info I've asked for in my previous comment. It will help all contributors understand what you're asking for, and that increases the possibility that someone will contribute a change.

**Comment by jrencz** [comment]
Ok, so summing things up:

I'd like those 3 examples 

```
(function foo(a,/* b,*/ c){})
(function foo(a, /*b, */c){})
(function foo(a, b/*, c*/){})
```

not to rise `Multiple spaces found after ','.` in rule `no-multi-spaces`. The first 2 now do. The third does not.

**Comment by gcochard** [comment]
I have seen this behavior too.

This will also trigger it:

```javascript
function foo(a,/* b, */c){}
```
`Multiple spaces found after ','. (no-multi-spaces)`

I can take a look at this over the weekend.

**Comment by gcochard** [comment]
Also, looking at your use case specifically @jrencz you might benefit from my [unparam](/gcochard/unparam) module regardless of this bug.

**Comment by gcochard** [comment]
These new tests are failing:

```javascript
(function(a,/* b, */c){})
(function(a,/*b,*/c){})
(function(a, /*b,*/c){})
(function(a,/*b,*/ c){})
(function(a, /*b,*/ c){})
```

These new tests are passing:

```javascript
(function(/*a, b, */c){})
(function(/*a, */b, c){})
(function(a, b/*, c*/){})
(function(a, b/*,c*/){})
(function(a, b /*,c*/){})
```

I'll have a PR shortly.

**Comment by jrencz** [comment]
@gcochard Thanks for the suggestion but unparam is not quite what will be useful for Angular DI pattern. I'd lie to have unused params detected statically. ESLint is probably a good tool to build such a detector but that's quite different case from the one I reported.

**Comment by michaelficarra** [comment]
I don't think this is worth "fixing". :-1:. To me, `no-multi-spaces` is not really about the space characters, but about the positioning.

**Comment by gcochard** [comment]
I disagree. I would say that commenting params out is a good thing. I have done this and Eslint has incorrectly flagged my code as having too many spaces when in reality there are none. Requirejs-style dependency injection is a good use case.

**Comment by michaelficarra** [comment]
A commented-out parameter should have been deleted. Comments are for humans, not for disabling code (or god forbid, having some magic effect). If you're only doing it temporarily during development, you should either not run a linter yet or ignore those linter warnings.

**Comment by gcochard** [comment]
It's still a bug. What if I were to put the expected type of parameter in the inline comment block? It's not in the jsdoc style, but it's still useful documentation.

**Comment by jrencz** [comment]
I haven't really thought about this before but it's even better reason to fix this than mine.

**Comment by nzakas** [comment]
Whether or not doing this is good coding practice really isn't relevant to this issue. The rule designed to detect multiple spaces is incorrectly reporting a comment as spaces. That's a bug in the rule that should be fixed.

**Comment by lo1tuma** [comment]
This bug is not limited to function parameters.

```js
 var a,/* b,*/c;

 var foo = [1,/* 2,*/3];

 var bar = {a: 1,/* b: 2*/c: 3};
```

```sh
foo.js
  1:5   error  Multiple spaces found after ','  no-multi-spaces
  3:12  error  Multiple spaces found after ','  no-multi-spaces
  5:15  error  Multiple spaces found after ','  no-multi-spaces
```

I’m not sure if the current implementation of this rule is a good approach to correctly detect all occurrences of multiple spaces (see #1603). Maybe we should rewrite this rule to fix both issues.

**Comment by nzakas** [comment]
I was thinking the same thing. Using tokens to do this seems like the wrong approach.

**Comment by gcochard** [comment]
Definitely, seems like using source lines might be less of a hack. One thing this does bring up though is whether multiple spaces inside of comments should be ignored.

I would vote yes on this, but it might be good to make it an option. I can refactor my pr today or tomorrow.

**Comment by nzakas** [comment]
Yeah, I think comments are an exception.

**Comment by doberkofler** [comment]
+1

**Comment by mourner** [comment]
Excited to see this fixed! Another common use case is commenting variables like this (e.g. for type annotation):

```js
var foo /* comment */  = 'bar';
```
